beetsplug/fetchart.py

```diff
@@ -1343,7 +1457,12 @@ def func(lib, opts, args):
 
     # Utilities converted from functions to methods on logging overhaul
 
-    def art_for_album(self, album, paths, local_only=False):
+    def art_for_album(
+        self,
+        album: Album,
+        paths: Optional[Sequence[bytes]],
+        local_only: bool = False,
+    ) -> Optional[Candidate]:
         """Given an Album object, returns a path to downloaded art for the
         album (or None if no art is found). If `maxwidth`, then images are
         resized to this maximum pixel size. If `quality` then resized images
```

Comment:
The thing is that `displayable_path` is called after `fetch_image`, which should always fetch the image and set `path`, which is then checked by `Candidate.validate`.

Encoding this in a way that the type checker can understood could be done using the _typestate pattern_ in other languages, but I don't think it can be done in Python. So, we'd probably need to add another `ValidatedCandidate` class that contains the `Candidate` and have `validate(self: Candidate) -> ValidatedCandidate`. I'm not sure that's worth the effort for now.

It is interesting that mypy doesn't flag this, though. These are typed functions, so it should check them by default...

Reply:
> The thing is that displayable_path is called after fetch_image, which should always fetch the image and set path, which is then checked by Candidate.validate.

I see what you mean - sounds like a very familiar situation 😁 In similar cases I end up strictly rejecting the input if it needs validating / is not ready for the task and delegate the validation responsibility to something else.